### ISSUE TITLE
What we want from the Scripting API when based on Simplified TD

### ISSUE THREAD

**Issue by mkovatsc** [w3c/wot-scripting-api#111]
The Thing Description (TD) uses a formal, and hence well-defined model that describes Things. Among other details, Things can have metadata attributes such as `name` and `id`, semantic metadata attributes such as `@context` and `@type` (obviously using JSON-LD keywords), interaction definitions in the structured attributes `properties`, `actions`, and `events`, and finally links to other Things or metadata documents in the structured attribute `links`.

This formal model and instances of the descriptions are the basis to instantiate software objects in a JavaScript runtime that act as object API to interact with the corresponding Thing. Since not all Things are the same, the object API needs to provide *introspection*, to check the values of the metadata attributes and which interactions are available. It would be most intuitive, when the software object has the same structure as the TD model. Most of this can easily be defined in WebIDL, as we have a list of identifiers.

**Problem 1:** the semantic attribute identifiers contain an `@` character, and hence are no valid identifiers for JavaScript. The solution should be to access them with the bracket notation, e.g., `myThing["@type"]` to introspect the semantic attribute `@type`. To my understanding, this would be done using a [`readonly maplike<key_type, value_type>;`](https://heycam.github.io/webidl/#idl-maplike) definition for the WebIDL `interface`. Others say "named getter" would do the trick. Please clarify. What we lose is the definition of which semantic attributes are to be expected (e.g., `@context` and `@type`). However, semantic attributes were made purely optional based on community feedback, so `maplike` should be fine, no?

Furthermore, the Thing Description must be extensible to be able to evolve other time and to adapt to the many different IoT ecosystems. For this, we allow domain-specific vocabularies, which can easily be imported in the TD document through the JSON-LD `@context` keyword. This would add additional semantic attributes that usually would be of the form `prefix:term` or could be a full IRI. The colon (`:`) also makes them invalid JavaScript identifiers, and hence we also need `maplike` for them.

**Problem 2:** There is a perceived risk of a security leak for some, as the WebIDL does not restrict what attributes will be part of the object API. Yet I fail to see the actual risk here: TDs are well defined and can be validated through multiple mechanisms such as JSON Schema, Shapes/SHACL, or even custom validators of the runtime that also knows which feature exactly is supported. The Scripting API can define a MUST be validated before instantiating the API object. Furthermore, the maplike mechanism defines which types are allowed: `key_type` is always a string, `value_type` is always a JSON data type -- never a function/callback/Promise/etc.

Summary:
* Does `maplike` enable introspection for the object API?
```javascript
interface ConsumedThing {
  readonly attribute DOMString name;
  /* business as usual continues... */
  /* introspection for extensible metadata attributes */
  readonly maplike<USVString, value_type>;
};
```
* How to best constrain the `value_type` to JSON data types only?
* Can there be a security risk when the API objects are generated from JSON documents that can further be validated against a formal model?

**Comment by zolkis** [comment]
WebIDL [restricts](https://heycam.github.io/webidl/#idl-maplike) using one `maplike` in a given interface and all its inherited interfaces (should be fine for this use case). Also, an interface containing `maplike` cannot be declared `iterable` (which may be a problem).

In principle I suggest separating JSON[-LD] _data_ from interface objects specified in the API doc.
The result of fetching a TD is _data_ that can be used for instantiating a standardized `ConsumedThing` object.
IMO the `ConsumedThing` object should be defined exactly and completely in the Scripting API spec.
For introspection it should be enough to have a method that returns a _data_ object/record that uses map-like access for semantic metadata attributes (only), or all attributes.

Spec prose can specify algorithms to validate `maplike` values as JSON data types.

The algorithms used in an API implementation (runtime) for validating a TD could be defined externally and referenced from the Scripting API spec.

**Comment by mkovatsc** [comment]
Maplike solves your first point:
> Maplike interfaces support an API for querying the map entries appropriate for the language binding. If the readonly keyword is not used, then it also supports an API for modifying the map entries.

> Note: In the ECMAScript language binding, the API for interacting with the map entries is similar to that available on ECMAScript Map objects. If the readonly keyword is used, this includes entries, forEach, get, has, keys, values, @@iterator methods, and a size getter. For read–write maplikes, it also includes clear, delete, and set methods.

The existence of `maplike` tells me we are wrong in what we want. I want to maximize user=developer-friendliness.

We do not need spec prose. We can cleanly define a type for `value_type`. I am just confused that there is no ready-to-use definition already.

**Comment by zolkis** [comment]
The problem is that based on the Web IDL spec the interfaces that contain a `maplike` are purely map-like, IOW you cannot make the `ConsumedThing` interface declare to be `maplike` AND also contain methods and other properties. The way I understand to do it is to define `ConsumedThing` as an API object, and for the data structure that reflects the TD (ThingDescription/ThingMetadata/etc) declare an interface to be maplike - but then all its properties are defined by mapping.

Also, IMHO it would be more clear (developer-friendly) to say
```javascript
let context = thing.td['@context'];
```
than to say
```javascript
let context = thing['@context'];
```
AFAIK [JSON types](https://heycam.github.io/webidl/#dfn-json-types) are denoted by `any`. We do need prose to specify/refer to externally defined algorithms on how to transform/validate values. 

We might also want to enumerate/refer to known (e.g. '@context', '@type' etc.) key-valuetype pairs for a TD, so that implementations could make a more thorough checks than key should be string and value type can be JSON type.

**Comment by mkovatsc** [comment]
> The problem is that based on the Web IDL spec the interfaces that contain a maplike are purely map-like

I cannot see how you come to this conclusion. There is no text on this. Yet there is text what member names must not be used:

> Maplike interfaces must not have any interface members named "entries", "forEach", "get", "has", "keys", "size", or "values", or have any inherited interfaces that have members with these names. 

A problem I see here is that with this, `myThing.properties.status.get()` will not work. Need a different name.

**Comment by mkovatsc** [comment]
Another aspect that came up is: who would RDF-style consumers of the TD want the annotations. @vcharpenay mentioned that in one of his explorations, he flattened the parsed JSON-LD object first to work with it. A ConsumedThing with readonly maplike access cannot be flattened; he would need a copy first anyway...

We need more reasons like this to decide whether we have the annotations inline with the API object or, for instance, get a copy via something like `myObject = myThing.getTD()`.

**Comment by mkovatsc** [comment]
> https://github.com/mozilla/gecko-dev/blob/250785729c6ba4fdae4d9967acf7c4733fb8b76f/dom/webidl/RTCStatsReport.webidl

```webidl
interface RTCStatsReport {
  readonly maplike<DOMString, object>;
  [ChromeOnly]
  readonly attribute DOMString mozPcid;
};
```

> https://github.com/mozilla/gecko-dev/blob/86897859913403b68829dbf9a154f5a87c4b0638/dom/webidl/TestInterfaceJSMaplikeSetlikeIterable.webidl

```webidl
interface TestInterfaceMaplike {
  maplike<DOMString, long>;
  void setInternal(DOMString aKey, long aValue);
  void clearInternal();
  boolean deleteInternal(DOMString aKey);
  boolean hasInternal(DOMString aKey);
};
```

**Comment by zolkis** [comment]
It seems fine to represent the type of semantic metadata with `USVString` in general. 

I agree that we cannot have an API that creates _interface_ objects based on a TD, but we can create dictionary objects (without functions) for representing them. The functions operating on that data should be in well defined interfaces. Validation must happen in the runtime, and we should be able to reference the algorithm from the Scripting API spec.

**Comment by zolkis** [comment]
Typically how a script would use semantic metadata? Any examples in mind?

**Comment by mkovatsc** [comment]
> e.g., `{"iot": "http://iotschema.org/"}`

Very good and obvious point I completely forgot!

I agree with @zolkis that we need concrete examples how a script would consume this. I would like to see something that combines simple scripting with the potential of Linked Data. This could work with through registered prefixes for WoT context files, e.g., so that `https://iot.schema.org/` always has the prefix `iot:`. Normal developers could rely on this to do simple string matching; Semantic Web developers still get the power of queries, reasoning, etc. @vcharpenay comments?

**Comment by mkovatsc** [comment]
My current assumption was for `ExposedThing` has been like this:

```javascript
myThing.properties.addProperty({ ... });
```

Would a named setter on the `ExposedThing` interface like `setter ThingProperty setProperty(string id, ThingProperty property)` allow the following:
```javascript
myThing.properties.status = { ... };
```
while the runtime internally calls `setProperty()`, which does a validation of `{ ... }`?

If so, we could have both, `addProperty()` and the named setter, which internally uses the implementation of `addProperty()`, no?

**Comment by zolkis** [comment]
> ```let semType = myThing.properties.status["@type"];
let unit = myThing.properties.status["iot:unit"];```

This doesn't seem very JS developer friendly to me (may be for semantic guys). It would be nice to [reduce to] use simple property names such as `"unit"` in scripts and keep namespace and context separately. 

Besides, what use cases do we have for mixing namespaces within an interaction?
I am just not convinced that operating in JSON-LD terms with JS syntax makes for a good API.
Let's base this on concrete use cases, examples and work it up from there.

**Comment by danielpeintner** [comment]
> This doesn't seem very JS developer friendly to me (may be for semantic guys). It would be nice to [reduce to] use simple property names such as "unit" in scripts and keep namespace and context separately.

This might not be very developer friendly but I guess we need to live with it one way or the other. JSON does not have a built-in namespace system such as XML and we need to differentiate `"iot:unit"` from `"gps:unit"` that might be used on the same node (_maybe not very likely but possible_).

The alternative would be to work on splitting name and namespace which does not sound very easy either.

Moreover, this points to another issue we run into when modifying a TD. Let's suppose we want to add semantic metadata with namespaces that have not been there before: we either **trust** the user to do the right thing by updating the `@context` first `{"foo": "http://foo.com/"}` before adding metadata such as `"foo:unit"` OR we provide an alternative way.

**Comment by mkovatsc** [comment]
> It would be nice to [reduce to] use simple property names such as "unit" 

@zolkis I guess using something intuitive such as `unit` distracted from the use case:

```javascript
let customInfo = myThing.properties.status["gms:COVParams"];
```

This is a mechnism to introspect any extension vocabulary. We can only have something simple as `unit` if we already know the term. This hold true for the TD core vocabulary. For the rest, we need something like this. @danielpeintner already pointed out the requirement for namespacing.

**Comment by mkovatsc** [comment]
> Moreover, this points to another issue we run into when modifying a TD

I drafted something for this in https://github.com/w3c/wot-scripting-api/issues/84

* Contexts and their prefixes must be provided when calling `WoT.produce()`
* Annotations are then limited to use the provided set of prefixes
* If the named setter calls a given function when setting (please see my question [above](https://github.com/w3c/wot-scripting-api/issues/111#issuecomment-383380705)), then this function could validate and throw an error if no valid prefix is used.

**Comment by zolkis** [comment]
Thank you @vcharpenay. I like the proposal - that is exactly what I asked on the F2F, whether could we have explicit getter/setter functions for semantic metadata. But for now we have agreed to use e.g.
`consumedThing["@type"]` rather than `consumedThing.getType()`, even though I personally like the latter more. The main reason for the former is extensibility beyond the core TD vocabulary.

We'll see into including the optional context objects.

**Comment by vcharpenay** [comment]
> To all: please provide use cases for unknown/extension objects in the TD -- and how they would be used by a script.

Geo-coordinates is an example that one might often encounter. In JSON-LD, they would be an object:

```js
let pos = system.getGPSCoordinates();
let p = new ThingProperty();
// http://schema.org/geo
p.coords = {
  // http://schema.org/longitude
  lon: pos[0],
  // http://schema.org/latitude
  lat: pos[1]
};
```

Another example is inline measurements for a device that does not store history data itself:

```js
let t = new ThingProperty();
t.setType('http://qudt.org/1.1/vocab/quantity/Temperature');
setInterval(() => {
  // http://www.w3.org/ns/sosa/madeObservation
  t.measurements.push({
    // http://www.w3.org/ns/sosa/hasSimpleResult
    value: system.getCurrentTemperatureValue(),
    // http://www.w3.org/ns/sosa/resultTime
    timestamp: Date.now().toISOString()
  });
}, 1000);
```

In general, any inline annotations that can be embedded in a TD saves bandwidth and may be of relevance in constrained networks. However, it does not imply that these annotations _must_ be exposed via the API.

**Comment by zolkis** [comment]
Fixed by #113.
